### PR TITLE
Viite 730 backend for split suravage link

### PR DESCRIPTION
**Pull request by tvenhola**
Backend operations for suravage link splitting in project mode:
Idempotent save operation (repeating operation overwrites old), removing the split.

### CODE REVIEW

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectAddressLinkBuilder.scala [hunk 1]
```diff
@@ -62,21 +62,27 @@ object ProjectAddressLinkBuilder extends AddressLinkBuilder {
...
                     discontinuity: Discontinuity,
                     startAddrMValue: Long, endAddrMValue: Long, startMValue: Double, endMValue: Double,
                     sideCode: SideCode, startCalibrationPoint: Option[CalibrationPoint], endCalibrationPoint: Option[CalibrationPoint],
-                    anomaly: Anomaly, lrmPositionId: Long, status: LinkStatus, roadAddressId : Long): ProjectAddressLink = {
+                    anomaly: Anomaly, lrmPositionId: Long, status: LinkStatus, roadAddressId: Long, connectedLinkId: Option[Long]): ProjectAddressLink = {
 
-    ProjectAddressLink(id, roadLink.linkId, geom,
+    val linkId =
+      if (connectedLinkId.nonEmpty && status == LinkStatus.New)
+        0L - roadLink.linkId
```
Comment (Campanha) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectAddressLinkBuilder.scala]: what will happen in this case . Will the linkId be a negative linkId?

Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectAddressLinkBuilder.scala]: Yes. There will be two links for the UI within a single link which will be represented by positive and negative numbers.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        GeometryUtils.areAdjacent(link1.geometry.head, link2.geometry.last, MaxDistanceForConnectedLinks)
+  }
+  private def suravageWithOptions(suravage: ProjectLink, templateLink: ProjectLink, split: SplitOptions, suravageM: Double,
+                              splitAddressM: Long) = {
+    (
+      suravage.copy(roadNumber = split.roadNumber,
+        roadPartNumber = split.roadPartNumber,
+        track = split.trackCode,
+        discontinuity = split.discontinuity,
+        //          ely = split.ely,
```
Comment (pedrosag) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala]: Can this ely code be removed?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        roadAddressId = templateLink.roadAddressId,
+        connectedLinkId = Some(templateLink.linkId),
+        geometry = GeometryUtils.truncateGeometry2D(suravage.geometry, 0.0, suravageM),
+        geometryLength = suravageM
+      ),
+      suravage.copy(roadNumber = split.roadNumber,
+        roadPartNumber = split.roadPartNumber,
+        track = split.trackCode,
+        discontinuity = split.discontinuity,
+        //          ely = split.ely,
```
Comment (pedrosag) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala]: Same as above, can this one be removed too?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        val dx = (p1 - p3).x
+        val normV2 = v2.normalize2D()
+        return Some(p3 + normV2.scale(dx/normV2.x))
+      }
+    } else if (Math.abs(v2.x) < 0.001) {
+      // second parameter is near vertical, switch places and rerun
+      return intersectionPoint((switchXY(p3), switchXY(p4)), (switchXY(p1),switchXY(p2))).map(switchXY)
+    }
+
+    val a = v1.y / v1.x
```
Comment (pedrosag) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/ProjectLinkSplitter.scala]: Please do review the variable names of this method (intersectionPoint if I am not mistaken) they are very non intuitive.

**Pull request merged**